modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java

```diff
@@ -0,0 +1,777 @@
+            Path filePath,
+            FileIoFactory ioFactory,
+            int pageSize
+    ) {
+        assert type == PageStore.TYPE_DATA || type == PageStore.TYPE_IDX : type;
+
+        this.type = type;
+        this.filePath = filePath;
+        this.ioFactory = ioFactory;
+        this.pageSize = pageSize;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void addWriteListener(PageWriteListener listener) {
+        listeners.add(listener);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void removeWriteListener(PageWriteListener listener) {
+        listeners.remove(listener);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void stop(boolean clean) throws IgniteInternalCheckedException {
+        try {
+            stop0(clean);
+        } catch (IOException e) {
+            throw new IgniteInternalCheckedException(
+                    "Failed to stop serving partition file [file=" + filePath + ", delete=" + clean + "]",
+                    e
+            );
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long allocatePage() throws IgniteInternalCheckedException {
+        init();
+
+        return allocatedBytes.getAndAdd(pageSize) / pageSize;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long pages() {
+        if (!initialized) {
+            return 0;
+        }
+
+        return allocatedBytes.get() / pageSize;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean read(long pageId, ByteBuffer pageBuf, boolean keepCrc) throws IgniteInternalCheckedException {
+        return read(pageId, pageBuf, !skipCrc, keepCrc);
+    }
+
+    /**
+     * Reads a page from the page store.
+     *
+     * @param pageId Page ID.
+     * @param pageBuf Page buffer to read into.
+     * @param checkCrc Check CRC on page.
+     * @param keepCrc By default reading zeroes CRC which was on file, but you can keep it in pageBuf if set keepCrc.
+     * @return {@code true} if page has been read successfully, {@code false} if page hasn't been written yet.
+     * @throws IgniteInternalCheckedException If reading failed (IO error occurred).
+     */
+    private boolean read(long pageId, ByteBuffer pageBuf, boolean checkCrc, boolean keepCrc) throws IgniteInternalCheckedException {
+        init();
+
+        try {
+            long off = pageOffset(pageId);
+
+            assert pageBuf.capacity() == pageSize;
+            assert pageBuf.remaining() == pageSize;
+            assert pageBuf.position() == 0;
+            assert pageBuf.order() == nativeOrder();
+            assert off <= allocatedBytes.get() : "calculatedOffset=" + off
+                    + ", allocated=" + allocatedBytes.get() + ", headerSize=" + headerSize() + ", filePath=" + filePath;
+
+            int n = readWithFailover(pageBuf, off);
+
+            // If page was not written yet, nothing to read.
+            if (n < 0) {
+                pageBuf.put(new byte[pageBuf.remaining()]);
```

Comment:
This is interesting, how can that happen? And why do we have to write zeroes

Reply:
-1 if the position is greater than or equal to the file's current size.
Removed nulling.

Reply:
In personal correspondence, we decided to leave it as it is, added some documentation.